File: backend/app/schemas/payment.py
"""
Kaasb Platform - Payment Schemas
"""

import uuid
from datetime import datetime

from pydantic import BaseModel, Field

# === Payment Account ===

class PaymentAccountSetup(BaseModel):
    """Setup a payment account."""
    provider: str = Field(pattern=r"^(qi_card|stripe|wise)$")
    # Wise-specific
    wise_email: str | None = Field(None, max_length=255)
    wise_currency: str = Field(default="USD", max_length=3)
    # Qi Card-specific (optional — used for account label)
    qi_card_phone: str | None = Field(None, max_length=20, description="Iraqi phone number linked to Qi Card")


class PaymentAccountResponse(BaseModel):
    id: uuid.UUID
    provider: str
    status: str
    external_account_id: str | None = None
    wise_email: str | None = None
    wise_currency: str = "USD"
    qi_card_phone: str | None = None
    is_default: bool = True
    verified_at: datetime | None = None
    created_at: datetime

    model_config = {"from_attributes": True}


# === Fund Escrow (Client) ===

class EscrowFundRequest(BaseModel):
    """Client funds escrow for a milestone."""
    milestone_id: uuid.UUID
    payment_method_id: str | None = None  # Stripe payment method (legacy)
    # callback_url and return_url are server-controlled constants — never user-supplied


class EscrowFundResponse(BaseModel):
    escrow_id: uuid.UUID
    milestone_id: uuid.UUID
    amount: float
    platform_fee: float
    freelancer_amount: float
    status: str
    # Qi Card payment redirect
    payment_redirect_url: str | None = None  # Redirect client here to complete payment
    qi_card_payment_id: str | None = None
    # Legacy Stripe
    client_secret: str | None = None
    message: str

    model_config = {"from_attributes": True}


# === Release Escrow ===

class EscrowReleaseResponse(BaseModel):
    escrow_id: uuid.UUID
    milestone_id: uuid.UUID
    amount: float
    freelancer_amount: float
    status: str
    message: str


# === Transaction ===

class TransactionResponse(BaseModel):
    id: uuid.UUID
    transaction_type: str
    status: str
    amount: float
    currency: str
    platform_fee: float
    net_amount: float
    description: str | None = None
    external_transaction_id: str | None = None
    completed_at: datetime | None = None
    created_at: datetime

    model_config = {"from_attributes": True}


class TransactionListResponse(BaseModel):
    transactions: list[TransactionResponse]
    total: int
    page: int
    page_size: int
    total_pages: int


# === Payment Dashboard Summary ===

class PaymentSummary(BaseModel):
    """User's payment overview."""
    total_earned: float = 0.0
    total_spent: float = 0.0
    pending_escrow: float = 0.0
    available_balance: float = 0.0
    total_platform_fees: float = 0.0
    transaction_count: int = 0
    payment_accounts: list[PaymentAccountResponse] = []


# === Qi Card Webhook ===

class QiCardWebhookEvent(BaseModel):
    """Qi Card webhook payload sent to our callback URL after payment."""
    payment_id: str
    order_id: str  # This is our escrow order_id
    status: str    # "completed" | "failed" | "cancelled"
    amount: int    # Amount in IQD
    merchant_id: str
    signature: str | None = None  # HMAC-SHA256 signature for verification


# === Legacy Stripe Webhook (kept for compatibility) ===

class StripeWebhookEvent(BaseModel):
    """Stripe webhook payload (partial — we parse what we need)."""
    id: str
    type: str
    data: dict


# === Payout Request (Freelancer) ===

class PayoutRequest(BaseModel):
    """Freelancer requests a payout."""
    amount: float = Field(ge=10.0, le=50000.0)
    payment_account_id: uuid.UUID


class PayoutResponse(BaseModel):
    transaction_id: uuid.UUID
    amount: float
    net_amount: float
    status: str
    provider: str
    message: str

File: frontend/src/types/payment.ts
// === Payment Types ===

export interface PaymentAccount {
  id: string;
  provider: string;
  status: string;
  external_account_id?: string;
  qi_card_phone?: string;
  is_default: boolean;
  verified_at?: string;
  created_at: string;
}

export interface TransactionDetail {
  id: string;
  transaction_type: string;
  status: string;
  amount: number;
  currency: string;
  platform_fee: number;
  net_amount: number;
  description?: string;
  external_transaction_id?: string;
  completed_at?: string;
  created_at: string;
}

export interface TransactionListResponse {
  transactions: TransactionDetail[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

export interface PaymentSummary {
  total_earned: number;
  total_spent: number;
  pending_escrow: number;
  available_balance: number;
  total_paid_out: number;
  pending_payout: number;
  total_platform_fees: number;
  transaction_count: number;
  payment_accounts: PaymentAccount[];
}

export interface EscrowFundResponse {
  escrow_id: string;
  milestone_id: string;
  amount: number;
  platform_fee: number;
  freelancer_amount: number;
  status: string;
  payment_redirect_url?: string;
  qi_card_payment_id?: string;
  message: string;
}

export interface PayoutResponse {
  transaction_id: string;
  amount: number;
  net_amount: number;
  status: string;
  provider: string;
  message: string;
}

// === Display helpers ===

export const TRANSACTION_TYPE_LABELS: Record<string, string> = {
  escrow_fund: "Escrow Funded",
  escrow_release: "Payment Released",
  escrow_refund: "Escrow Refunded",
  platform_fee: "Platform Fee",
  payout: "Payout",
};

export const TRANSACTION_STATUS_COLORS: Record<string, string> = {
  pending: "bg-yellow-50 text-yellow-700",
  processing: "bg-blue-50 text-blue-700",
  completed: "bg-green-50 text-green-700",
  failed: "bg-red-50 text-red-700",
  refunded: "bg-gray-100 text-gray-600",
  cancelled: "bg-gray-100 text-gray-500",
};

export const PROVIDER_LABELS: Record<string, string> = {
  qi_card: "Qi Card",
  manual: "Manual",
};
